File: Backend/EcoBackend.API/Services/ActivityService.cs
using EcoBackend.Core.Entities;
using EcoBackend.Infrastructure.Data;
using EcoBackend.API.DTOs;
using Microsoft.EntityFrameworkCore;

namespace EcoBackend.API.Services;

public class ActivityService
{
    private readonly EcoDbContext _context;

    public ActivityService(EcoDbContext context)
    {
        _context = context;
    }

    // ========== Categories ==========

    public async Task<List<ActivityCategoryDto>> GetCategoriesAsync()
    {
        var categories = await _context.ActivityCategories
            .Include(c => c.ActivityTypes)
            .ToListAsync();

        return categories.Select(c => new ActivityCategoryDto
        {
            Id = c.Id,
            Name = c.Name,
            Icon = c.Icon,
            Color = c.Color,
            Description = c.Description,
            ActivityTypes = c.ActivityTypes.Select(at => new ActivityTypeDto
            {
                Id = at.Id,
                CategoryId = at.CategoryId,
                Name = at.Name,
                Icon = at.Icon,
                CO2Impact = at.CO2Impact,
                ImpactUnit = at.ImpactUnit,
                IsEcoFriendly = at.IsEcoFriendly,
                Points = at.Points
            }).ToList()
        }).ToList();
    }

    public async Task<ActivityCategoryDto?> GetCategoryByIdAsync(int id)
    {
        var category = await _context.ActivityCategories
            .Include(c => c.ActivityTypes)
            .FirstOrDefaultAsync(c => c.Id == id);

        if (category == null) return null;

        return new ActivityCategoryDto
        {
            Id = category.Id,
            Name = category.Name,
            Icon = category.Icon,
            Color = category.Color,
            Description = category.Description,
            ActivityTypes = category.ActivityTypes.Select(at => new ActivityTypeDto
            {
                Id = at.Id,
                CategoryId = at.CategoryId,
                Name = at.Name,
                Icon = at.Icon,
                CO2Impact = at.CO2Impact,
                ImpactUnit = at.ImpactUnit,
                IsEcoFriendly = at.IsEcoFriendly,
                Points = at.Points
            }).ToList()
        };
    }

    // ========== Activity Types ==========

    public async Task<List<ActivityTypeDto>> GetActivityTypesAsync(int? categoryId)
    {
        var query = _context.ActivityTypes.AsQueryable();

        if (categoryId.HasValue)
            query = query.Where(at => at.CategoryId == categoryId.Value);

        var types = await query.ToListAsync();

        return types.Select(at => new ActivityTypeDto
        {
            Id = at.Id,
            CategoryId = at.CategoryId,
            Name = at.Name,
            Icon = at.Icon,
            CO2Impact = at.CO2Impact,
            ImpactUnit = at.ImpactUnit,
            IsEcoFriendly = at.IsEcoFriendly,
            Points = at.Points
        }).ToList();
    }

    public async Task<ActivityTypeDto?> GetActivityTypeByIdAsync(int id)
    {
        var activityType = await _context.ActivityTypes
            .Include(at => at.Category)
            .FirstOrDefaultAsync(at => at.Id == id);

        if (activityType == null) return null;

        return new ActivityTypeDto
        {
            Id = activityType.Id,
            CategoryId = activityType.CategoryId,
            Name = activityType.Name,
            Icon = activityType.Icon,
            CO2Impact = activityType.CO2Impact,
            ImpactUnit = activityType.ImpactUnit,
            IsEcoFriendly = activityType.IsEcoFriendly,
            Points = activityType.Points
        };
    }

    // ========== Activities CRUD ==========

    public async Task<List<ActivityDto>> GetActivitiesAsync(int userId, DateTime? startDate, DateTime? endDate, int? category)
    {
        var query = _context.Activities
            .Include(a => a.ActivityType)
            .ThenInclude(at => at.Category)
            .Where(a => a.UserId == userId);

        if (startDate.HasValue)
            query = query.Where(a => a.ActivityDate >= startDate.Value);
        if (endDate.HasValue)
            query = query.Where(a => a.ActivityDate <= endDate.Value);
        if (category.HasValue)
            query = query.Where(a => a.ActivityType.CategoryId == category.Value);

        var activities = await query.OrderByDescending(a => a.ActivityDate).ToListAsync();

        return activities.Select(MapToActivityDto).ToList();
    }

    public async Task<ActivityDto?> GetActivityByIdAsync(int id, int userId)
    {
        var activity = await _context.Activities
            .Include(a => a.ActivityType)
            .ThenInclude(at => at.Category)
            .FirstOrDefaultAsync(a => a.Id == id && a.UserId == userId);

        if (activity == null) return null;

        return MapToActivityDto(activity);
    }

    public async Task<ActivityDto> CreateActivityAsync(int userId, ActivityCreateDto dto)
    {
        var activityType = await _context.ActivityTypes.FindAsync(dto.ActivityTypeId);
        if (activityType == null)
            throw new ArgumentException("Invalid activity type");

        var activity = new Activity
        {
            UserId = userId,
            ActivityTypeId = dto.ActivityTypeId,
            Quantity = dto.Quantity,
            Unit = dto.Unit ?? "",
            Notes = dto.Notes ?? "",
            Latitude = dto.Latitude,
            Longitude = dto.Longitude,
            LocationName = dto.LocationName ?? "",
            ActivityDate = dto.ActivityDate,
            ActivityTime = dto.ActivityTime,
            CO2Impact = activityType.CO2Impact * dto.Quantity,
            PointsEarned = activityType.IsEcoFriendly ? (int)(activityType.Points * dto.Quantity) : 0
        };

        _context.Activities.Add(activity);
        await _context.SaveChangesAsync();

        // Update user stats
        await UpdateUserStatsAsync(userId, activity);

        activity = await _context.Activities
            .Include(a => a.ActivityType)
            .ThenInclude(at => at.Category)
            .FirstAsync(a => a.Id == activity.Id);

        return MapToActivityDto(activity);
    }

    public async Task<bool> DeleteActivityAsync(int id, int userId)
    {
        var activity = await _context.Activities
            .FirstOrDefaultAsync(a => a.Id == id && a.UserId == userId);

        if (activity == null) return false;

        _context.Activities.Remove(activity);
        await _context.SaveChangesAsync();
        return true;
    }

    // ========== Today & Summary ==========

    public async Task<List<ActivityDto>> GetTodayActivitiesAsync(int userId)
    {
        var today = DateTime.UtcNow.Date;

        var activities = await _context.Activities
            .Include(a => a.ActivityType)
            .ThenInclude(at => at.Category)
            .Where(a => a.UserId == userId && a.ActivityDate == today)
            .OrderByDescending(a => a.CreatedAt)
            .ToListAsync();

        return activities.Select(MapToActivityDto).ToList();
    }

    public async Task<ActivitySummaryDto> GetSummaryAsync(int userId, int days)
    {
        var endDate = DateTime.UtcNow.Date;
        var startDate = days == 1 ? endDate : endDate.AddDays(-(days - 1));

        var activities = await _context.Activities
            .Include(a => a.ActivityType)
            .ThenInclude(at => at!.Category)
            .Where(a => a.UserId == userId && a.ActivityDate >= startDate && a.ActivityDate <= endDate)
            .ToListAsync();

        var totalActivities = activities.Count;
        var totalPoints = activities.Sum(a => a.PointsEarned);
        var totalCO2Saved = Math.Abs(activities.Where(a => a.CO2Impact < 0).Sum(a => a.CO2Impact));
        var totalCO2Emitted = activities.Where(a => a.CO2Impact > 0).Sum(a => a.CO2Impact);

        var byCategory = activities
            .Where(a => a.ActivityType?.Category != null)
            .GroupBy(a => a.ActivityType!.Category!.Name)
            .ToDictionary(
                g => g.Key,
                g => new CategoryStats
                {
                    Count = g.Count(),
                    CO2Impact = g.Sum(a => a.CO2Impact)
                }
            );

        return new ActivitySummaryDto
        {
            StartDate = startDate,
            EndDate = endDate,
            TotalActivities = totalActivities,
            TotalPoints = totalPoints,
            TotalCO2Saved = totalCO2Saved,
            TotalCO2Emitted = totalCO2Emitted,
            ByCategory = byCategory
        };
    }

    // ========== Tips ==========

    public async Task<List<TipDto>> GetTipsAsync(int? category)
    {
        var query = _context.Tips.Where(t => t.IsActive).AsQueryable();

        if (category.HasValue)
            query = query.Where(t => t.CategoryId == category.Value);

        var tips = await query.OrderByDescending(t => t.Priority).ToListAsync();

        return tips.Select(t => new TipDto
        {
            Id = t.Id,
            Title = t.Title,
            Content = t.Content,
            ImpactDescription = t.ImpactDescription,
            Priority = t.Priority
        }).ToList();
    }

    public async Task<TipDto?> GetTipByIdAsync(int id)
    {
        var tip = await _context.Tips.FirstOrDefaultAsync(t => t.Id == id);

        if (tip == null) return null;

        return new TipDto
        {
            Id = tip.Id,
            Title = tip.Title,
            Content = tip.Content,
            ImpactDescription = tip.ImpactDescription,
            Priority = tip.Priority
        };
    }

    public async Task<TipDto> GetDailyTipAsync()
    {
        var tips = await _context.Tips
            .Where(t => t.IsActive)
            .ToListAsync();

        if (!tips.Any())
        {
            return new TipDto
            {
                Id = 0,
                Title = "Keep Going!",
                Content = "Every small action counts towards a greener future.",
                ImpactDescription = "Your efforts make a difference!",
                Priority = 1
            };
        }

        var random = new Random();
        var tip = tips[random.Next(tips.Count)];

        return new TipDto
        {
            Id = tip.Id,
            Title = tip.Title,
            Content = tip.Content,
            ImpactDescription = tip.ImpactDescription,
            Priority = tip.Priority
        };
    }

    // ========== Activity History ==========

    public async Task<Dictionary<string, List<object>>> GetActivityHistoryAsync(int userId, int days)
    {
        var startDate = DateTime.UtcNow.Date.AddDays(-days);

        var activities = await _context.Activities
            .Where(a => a.UserId == userId && a.ActivityDate >= startDate)
            .Include(a => a.ActivityType)
            .ThenInclude(at => at.Category)
            .OrderByDescending(a => a.ActivityDate)
            .ThenByDescending(a => a.CreatedAt)
            .ToListAsync();

        return activities
            .GroupBy(a => a.ActivityDate.ToString("yyyy-MM-dd"))
            .ToDictionary(
                g => g.Key,
                g => g.Select(a => (object)new
                {
                    a.Id,
                    ActivityTypeId = a.ActivityTypeId,
                    ActivityTypeName = a.ActivityType.Name,
                    CategoryName = a.ActivityType.Category?.Name ?? "",
                    a.Quantity,
                    a.Unit,
                    a.Notes,
                    CO2Impact = a.CO2Impact,
                    PointsEarned = a.PointsEarned,
                    a.Latitude,
                    a.Longitude,
                    a.LocationName,
                    ActivityDate = a.ActivityDate.ToString("yyyy-MM-dd"),
                    a.ActivityTime,
                    a.IsAutoDetected,
                    a.CreatedAt
                }).ToList()
            );
    }

    // ========== Private Helpers ==========

    private async Task UpdateUserStatsAsync(int userId, Activity activity)
    {
        var user = await _context.Users.FindAsync(userId);
        if (user == null) return;

        var dailyScore = await _context.DailyScores
            .FirstOrDefaultAsync(ds => ds.UserId == userId && ds.Date == activity.ActivityDate);

        if (dailyScore == null)
        {
            dailyScore = new DailyScore
            {
                UserId = userId,
                Date = activity.ActivityDate
            };
            _context.DailyScores.Add(dailyScore);
        }

        var dailyActivities = await _context.Activities
            .Where(a => a.UserId == userId && a.ActivityDate == activity.ActivityDate)
            .ToListAsync();

        var co2Saved = Math.Abs(dailyActivities.Where(a => a.CO2Impact < 0).Sum(a => a.CO2Impact));
        var co2Emitted = dailyActivities.Where(a => a.CO2Impact > 0).Sum(a => a.CO2Impact);

        dailyScore.CO2Saved = co2Saved;
        dailyScore.CO2Emitted = co2Emitted;
        dailyScore.Score = dailyActivities.Sum(a => a.PointsEarned);

        var totalSaved = Math.Abs(await _context.Activities
            .Where(a => a.UserId == userId && a.CO2Impact < 0)
            .SumAsync(a => a.CO2Impact));

        user.TotalCO2Saved = totalSaved;
        user.ExperiencePoints += activity.PointsEarned;
        user.CalculateLevel();

        await _context.SaveChangesAsync();
    }

    private ActivityDto MapToActivityDto(Activity activity)
    {
        return new ActivityDto
        {
            Id = activity.Id,
            UserId = activity.UserId,
            ActivityTypeId = activity.ActivityTypeId,
            Quantity = activity.Quantity,
            Unit = activity.Unit,
            Notes = activity.Notes,
            CO2Impact = activity.CO2Impact,
            PointsEarned = activity.PointsEarned,
            Latitude = activity.Latitude,
            Longitude = activity.Longitude,
            LocationName = activity.LocationName,
            ActivityDate = activity.ActivityDate,
            ActivityTime = activity.ActivityTime,
            IsAutoDetected = activity.IsAutoDetected,
            ActivityType = activity.ActivityType == null ? null! : new ActivityTypeDto
            {
                Id = activity.ActivityType.Id,
                CategoryId = activity.ActivityType.CategoryId,
                Name = activity.ActivityType.Name,
                Icon = activity.ActivityType.Icon,
                CO2Impact = activity.ActivityType.CO2Impact,
                ImpactUnit = activity.ActivityType.ImpactUnit,
                IsEcoFriendly = activity.ActivityType.IsEcoFriendly,
                Points = activity.ActivityType.Points,
                Category = activity.ActivityType.Category == null ? null! : new ActivityCategoryDto
                {
                    Id = activity.ActivityType.Category.Id,
                    Name = activity.ActivityType.Category.Name,
                    Icon = activity.ActivityType.Category.Icon,
                    Color = activity.ActivityType.Category.Color,
                    Description = activity.ActivityType.Category.Description
                }
            }
        };
    }
}

File: Frontend/lib/services/activity_service.dart
import 'package:dio/dio.dart';
import '../core/config/api_config.dart';
import '../core/di/service_locator.dart';
import '../core/network/dio_client.dart';

/// Model for activity category
class ActivityCategory {
  final int id;
  final String name;
  final String icon;
  final String color;
  final String description;
  final List<ActivityType> activityTypes;

  ActivityCategory({
    required this.id,
    required this.name,
    required this.icon,
    required this.color,
    required this.description,
    required this.activityTypes,
  });

  factory ActivityCategory.fromJson(Map<String, dynamic> json) {
    final typesJson = json['activity_types'] as List<dynamic>? ?? [];
    return ActivityCategory(
      id: json['id'] ?? 0,
      name: json['name'] ?? '',
      icon: json['icon'] ?? '',
      color: json['color'] ?? '#4CAF50',
      description: json['description'] ?? '',
      activityTypes: typesJson.map((t) => ActivityType.fromJson(t)).toList(),
    );
  }
}

/// Model for activity type
class ActivityType {
  final int id;
  final String name;
  final String icon;
  final double co2Impact;
  final String impactUnit;
  final bool isEcoFriendly;
  final int points;
  final int? categoryId;
  final String? categoryName;

  ActivityType({
    required this.id,
    required this.name,
    required this.icon,
    required this.co2Impact,
    required this.impactUnit,
    required this.isEcoFriendly,
    required this.points,
    this.categoryId,
    this.categoryName,
  });

  factory ActivityType.fromJson(Map<String, dynamic> json) {
    return ActivityType(
      id: json['id'] ?? 0,
      name: json['name'] ?? '',
      icon: json['icon'] ?? '',
      co2Impact: (json['co2_impact'] ?? 0).toDouble(),
      impactUnit: json['impact_unit'] ?? 'per instance',
      isEcoFriendly: json['is_eco_friendly'] ?? false,
      points: json['points'] ?? 0,
      categoryId: json['category'],
      categoryName: json['category_name'],
    );
  }
}

/// Model for logged activity
class Activity {
  final int id;
  final int activityTypeId;
  final String activityTypeName;
  final String categoryName;
  final double quantity;
  final String unit;
  final String notes;
  final double co2Impact;
  final int pointsEarned;
  final double? latitude;
  final double? longitude;
  final String? locationName;
  final String activityDate;
  final String? activityTime;
  final bool isAutoDetected;
  final String createdAt;

  Activity({
    required this.id,
    required this.activityTypeId,
    required this.activityTypeName,
    required this.categoryName,
    required this.quantity,
    required this.unit,
    required this.notes,
    required this.co2Impact,
    required this.pointsEarned,
    this.latitude,
    this.longitude,
    this.locationName,
    required this.activityDate,
    this.activityTime,
    required this.isAutoDetected,
    required this.createdAt,
  });

  factory Activity.fromJson(Map<String, dynamic> json) {
    return Activity(
      id: json['id'] ?? 0,
      activityTypeId: json['activity_type'] ?? 0,
      activityTypeName: json['activity_type_name'] ?? '',
      categoryName: json['category_name'] ?? '',
      quantity: (json['quantity'] ?? 1).toDouble(),
      unit: json['unit'] ?? '',
      notes: json['notes'] ?? '',
      co2Impact: (json['co2_impact'] ?? 0).toDouble(),
      pointsEarned: json['points_earned'] ?? 0,
      latitude: json['latitude']?.toDouble(),
      longitude: json['longitude']?.toDouble(),
      locationName: json['location_name'],
      activityDate: json['activity_date'] ?? '',
      activityTime: json['activity_time'],
      isAutoDetected: json['is_auto_detected'] ?? false,
      createdAt: json['created_at'] ?? '',
    );
  }
}

/// Model for activity summary
class ActivitySummary {
  final String startDate;
  final String endDate;
  final int totalActivities;
  final int totalPoints;
  final double totalCo2Saved;
  final double totalCo2Emitted;
  final Map<String, dynamic> byCategory;

  ActivitySummary({
    required this.startDate,
    required this.endDate,
    required this.totalActivities,
    required this.totalPoints,
    required this.totalCo2Saved,
    required this.totalCo2Emitted,
    required this.byCategory,
  });

  factory ActivitySummary.fromJson(Map<String, dynamic> json) {
    return ActivitySummary(
      startDate: json['start_date']?.toString() ?? '',
      endDate: json['end_date']?.toString() ?? '',
      totalActivities: json['total_activities'] ?? 0,
      totalPoints: json['total_points'] ?? 0,
      totalCo2Saved: (json['total_co2_saved'] ?? 0).toDouble(),
      totalCo2Emitted: (json['total_co2_emitted'] ?? 0).toDouble(),
      byCategory: json['by_category'] ?? {},
    );
  }
}

/// Model for daily tip
class Tip {
  final int id;
  final int? categoryId;
  final String? categoryName;
  final String title;
  final String content;
  final String impactDescription;

  Tip({
    required this.id,
    this.categoryId,
    this.categoryName,
    required this.title,
    required this.content,
    required this.impactDescription,
  });

  factory Tip.fromJson(Map<String, dynamic> json) {
    return Tip(
      id: json['id'] ?? 0,
      categoryId: json['category'],
      categoryName: json['category_name'],
      title: json['title'] ?? '',
      content: json['content'] ?? '',
      impactDescription: json['impact_description'] ?? '',
    );
  }
}

/// Service for handling activity-related API calls
class ActivityService {
  static final _dio = sl<DioClient>().dio;
  static const String _baseUrl = ApiConfig.activitiesUrl;

  /// Get all activity categories with their types
  static Future<List<ActivityCategory>> getCategories() async {
    try {
      final response = await _dio.get('$_baseUrl/categories');

      if (response.data == null) return [];
      final decoded = response.data;
      final List<dynamic> data = decoded is List ? decoded : (decoded is Map ? (decoded['results'] ?? []) : []);
      return data.map((c) => ActivityCategory.fromJson(c as Map<String, dynamic>)).toList();
    } on DioException catch (e) {
      if (e.response?.statusCode == 401) {
        throw ActivityException(
          'Authentication failed. Please log in again.',
          statusCode: 401,
        );
      }
      throw ActivityException(
        'Failed to load categories. ${e.message}',
        statusCode: e.response?.statusCode,
      );
    } catch (e) {
      if (e is ActivityException) rethrow;
      throw ActivityException('Failed to load categories: ${e.toString()}');
    }
  }

  /// Get activity types, optionally filtered by category
  static Future<List<ActivityType>> getActivityTypes({int? categoryId}) async {
    try {
      final url = categoryId != null
          ? '$_baseUrl/types?category=$categoryId'
          : '$_baseUrl/types';
      final response = await _dio.get(url);

      final decoded = response.data;
      final List<dynamic> data = decoded is List ? decoded : (decoded['results'] ?? []);
      return data.map((t) => ActivityType.fromJson(t)).toList();
    } on DioException catch (e) {
      if (e.response?.statusCode == 401) {
        throw ActivityException(
          'Authentication failed. Please log in again.',
          statusCode: 401,
        );
      }
      throw ActivityException(
        'Failed to load activity types. ${e.message}',
        statusCode: e.response?.statusCode,
      );
    } catch (e) {
      if (e is ActivityException) rethrow;
      throw ActivityException('Failed to load activity types: ${e.toString()}');
    }
  }

  /// Log a new activity
  static Future<Activity> logActivity({
    required int activityTypeId,
    required double quantity,
    String? unit,
    String? notes,
    double? latitude,
    double? longitude,
    String? locationName,
    required String activityDate,
    String? activityTime,
    bool isAutoDetected = false,
  }) async {
    try {
      final body = <String, dynamic>{
        'activity_type': activityTypeId,
        'quantity': quantity,
        'activity_date': activityDate,
        'is_auto_detected': isAutoDetected,
      };

      if (unit != null && unit.isNotEmpty) body['unit'] = unit;
      if (notes != null && notes.isNotEmpty) body['notes'] = notes;
      if (latitude != null) body['latitude'] = latitude;
      if (longitude != null) body['longitude'] = longitude;
      if (locationName != null && locationName.isNotEmpty) {
        body['location_name'] = locationName;
      }
      if (activityTime != null && activityTime.isNotEmpty) {
        body['activity_time'] = activityTime;
      }

      final response = await _dio.post(_baseUrl, data: body);

      if (response.statusCode == 201) {
        return Activity.fromJson(response.data);
      }
      throw ActivityException(
        'Failed to log activity. Server returned ${response.statusCode}',
        statusCode: response.statusCode,
      );
    } on DioException catch (e) {
      if (e.response?.statusCode == 401) {
        throw ActivityException(
          'Authentication failed. Please log in again.',
          statusCode: 401,
        );
      } else if (e.response?.statusCode == 400) {
        String message = 'Invalid activity data.';
        final error = e.response?.data;
        if (error is Map) {
          message = error.entries.map((e) => '${e.key}: ${e.value}').join(', ');
        }
        throw ActivityException(message, statusCode: 400);
      }
      throw ActivityException(
        'Failed to log activity. ${e.message}',
        statusCode: e.response?.statusCode,
      );
    } catch (e) {
      if (e is ActivityException) rethrow;
      throw ActivityException('Failed to log activity: ${e.toString()}');
    }
  }

  /// Get today's activities
  static Future<List<Activity>> getTodayActivities() async {
    try {
      final response = await _dio.get('$_baseUrl/today');

      if (response.data == null) return [];
      final List<dynamic> data = response.data is List ? response.data : [];
      return data.map((a) => Activity.fromJson(a as Map<String, dynamic>)).toList();
    } on DioException catch (e) {
      if (e.response?.statusCode == 401) {
        throw ActivityException(
          'Authentication failed. Please log in again.',
          statusCode: 401,
        );
      }
      throw ActivityException(
        'Failed to load today\'s activities. ${e.message}',
        statusCode: e.response?.statusCode,
      );
    } catch (e) {
      if (e is ActivityException) rethrow;
      throw ActivityException('Failed to load activities: ${e.toString()}');
    }
  }

  /// Get activity summary for a date range
  static Future<ActivitySummary> getSummary({int days = 7}) async {
    try {
      final response = await _dio.get('$_baseUrl/summary?days=$days');

      if (response.data == null || response.data is! Map) {
        return ActivitySummary(
          startDate: '', endDate: '', totalActivities: 0,
          totalPoints: 0, totalCo2Saved: 0, totalCo2Emitted: 0, byCategory: {},
        );
      }
      return ActivitySummary.fromJson(response.data as Map<String, dynamic>);
    } on DioException catch (e) {
      if (e.response?.statusCode == 401) {
        throw ActivityException(
          'Authentication failed. Please log in again.',
          statusCode: 401,
        );
      }
      throw ActivityException(
        'Failed to load activity summary. ${e.message}',
        statusCode: e.response?.statusCode,
      );
    } catch (e) {
      if (e is ActivityException) rethrow;
      throw ActivityException('Failed to load summary: ${e.toString()}');
    }
  }

  /// Get activity history grouped by date
  static Future<Map<String, List<Activity>>> getHistory({int days = 30}) async {
    try {
      final response = await _dio.get('$_baseUrl/history?days=$days');

      if (response.data == null || response.data is! Map) return {};
      final Map<String, dynamic> data = Map<String, dynamic>.from(response.data);
      final result = <String, List<Activity>>{};

      data.forEach((date, activities) {
        if (activities is List) {
          result[date] = activities.map((a) => Activity.fromJson(a as Map<String, dynamic>)).toList();
        }
      });

      return result;
    } on DioException catch (e) {
      if (e.response?.statusCode == 401) {
        throw ActivityException(
          'Authentication failed. Please log in again.',
          statusCode: 401,
        );
      }
      throw ActivityException(
        'Failed to load activity history. ${e.message}',
        statusCode: e.response?.statusCode,
      );
    } catch (e) {
      if (e is ActivityException) rethrow;
      throw ActivityException('Failed to load history: ${e.toString()}');
    }
  }

  /// Get daily tip
  static Future<Tip?> getDailyTip() async {
    try {
      final response = await _dio.get('$_baseUrl/tips/daily');

      return Tip.fromJson(response.data);
    } catch (e) {
      return null;
    }
  }

  /// Delete an activity
  static Future<void> deleteActivity(int activityId) async {
    try {
      final response = await _dio.delete('$_baseUrl/$activityId');

      if (response.statusCode != 204 && response.statusCode != 200) {
        throw ActivityException(
          'Failed to delete activity.',
          statusCode: response.statusCode,
        );
      }
    } on DioException catch (e) {
      if (e.response?.statusCode == 401) {
        throw ActivityException(
          'Authentication failed. Please log in again.',
          statusCode: 401,
        );
      } else if (e.response?.statusCode == 404) {
        throw ActivityException(
          'Activity not found.',
          statusCode: 404,
        );
      }
      throw ActivityException(
        'Failed to delete activity. ${e.message}',
        statusCode: e.response?.statusCode,
      );
    } catch (e) {
      if (e is ActivityException) rethrow;
      throw ActivityException('Failed to delete activity: ${e.toString()}');
    }
  }

  /// Search activities by query
  static Future<List<Map<String, dynamic>>> searchActivities(String query) async {
    try {
      final response = await _dio.get('$_baseUrl?search=$query');

      final List<dynamic> data = response.data;
      return data.map((item) => item as Map<String, dynamic>).toList();
    } on DioException catch (e) {
      throw ActivityException(
        'Search failed. ${e.message}',
        statusCode: e.response?.statusCode,
      );
    } catch (e) {
      if (e is ActivityException) rethrow;
      throw ActivityException('Search failed: ${e.toString()}');
    }
  }
}

/// Custom exception for activity errors
class ActivityException implements Exception {
  final String message;
  final int? statusCode;

  ActivityException(this.message, {this.statusCode});

  @override
  String toString() => message;
}
